File: WriteBuffer.cpp
#include "stdafx.h"
#include "WriteBuffer.h"

using namespace System;
using namespace System::IO;
using namespace Microsoft::Windows::ProjFS;

WriteBuffer::WriteBuffer(
    unsigned long bufferSize,
    unsigned long alignment)
{
    this->buffer = (unsigned char*) _aligned_malloc(bufferSize, alignment);
    if (this->buffer == nullptr)
    {
        throw gcnew OutOfMemoryException("Unable to allocate WriteBuffer");
    }

    this->namespaceCtx = nullptr;
    this->stream = gcnew UnmanagedMemoryStream(buffer, bufferSize, bufferSize, FileAccess::Write);
}

WriteBuffer::WriteBuffer(
    unsigned long bufferSize,
    PRJ_NAMESPACE_VIRTUALIZATION_CONTEXT namespaceCtx,
    ApiHelper^ apiHelper)
{
    this->buffer = (unsigned char*) apiHelper->_PrjAllocateAlignedBuffer(namespaceCtx, bufferSize);
    if (this->buffer == nullptr)
    {
        throw gcnew OutOfMemoryException("Unable to allocate WriteBuffer");
    }

    this->namespaceCtx = namespaceCtx;
    this->apiHelper = apiHelper;
    this->stream = gcnew UnmanagedMemoryStream(buffer, bufferSize, bufferSize, FileAccess::Write);
}

WriteBuffer::~WriteBuffer()
{
    delete this->stream;
    this->!WriteBuffer();
}

WriteBuffer::!WriteBuffer()
{
    if (this->namespaceCtx)
    {
        this->apiHelper->_PrjFreeAlignedBuffer(this->buffer);
    }
    else
    {
        _aligned_free(this->buffer);
    }
}

long long WriteBuffer::Length::get(void)
{
    return this->stream->Length;
}

UnmanagedMemoryStream^ WriteBuffer::Stream::get(void)
{
    return this->stream;
}

IntPtr WriteBuffer::Pointer::get(void)
{
    return IntPtr(this->buffer);
}

File: ProjectedFSLib.Managed/WriteBuffer.cs
#nullable disable
using System;
using System.IO;
using System.Runtime.InteropServices;

namespace Microsoft.Windows.ProjFS
{
    /// <summary>
    /// Pure C# P/Invoke implementation of IWriteBuffer, using PrjAllocateAlignedBuffer/PrjFreeAlignedBuffer.
    /// Replaces the C++/CLI WriteBuffer from ProjectedFSLib.Managed.dll for NativeAOT compatibility.
    /// </summary>
    public class WriteBuffer : IWriteBuffer
    {
        private IntPtr _buffer;
        private bool _disposed;

        internal unsafe WriteBuffer(SafeProjFsHandle virtualizationContext, uint desiredBufferSize)
        {
            _buffer = ProjFSNative.PrjAllocateAlignedBuffer(virtualizationContext, new UIntPtr(desiredBufferSize));
            if (_buffer == IntPtr.Zero)
            {
                throw new InvalidOperationException("PrjAllocateAlignedBuffer returned null — insufficient memory or invalid context.");
            }

            Length = desiredBufferSize;
            byte* pBuf = (byte*)_buffer.ToPointer();
            Stream = new UnmanagedMemoryStream(pBuf, desiredBufferSize, desiredBufferSize, FileAccess.Write);
        }

        ~WriteBuffer()
        {
            Dispose(false);
        }

#pragma warning disable CA1720 // Identifier contains type name — established public API
        public IntPtr Pointer => _buffer;
#pragma warning restore CA1720

        public UnmanagedMemoryStream Stream { get; private set; }

        public long Length { get; }

        public void Dispose()
        {
            Dispose(true);
            GC.SuppressFinalize(this);
        }

        protected void Dispose(bool disposing)
        {
            if (_disposed)
            {
                return;
            }

            if (disposing)
            {
                Stream?.Dispose();
                Stream = null;
            }

            if (_buffer != IntPtr.Zero)
            {
                ProjFSNative.PrjFreeAlignedBuffer(_buffer);
                _buffer = IntPtr.Zero;
            }

            _disposed = true;
        }
    }
}
